src/test/groovy/graphql/normalized/ExecutableNormalizedOperationFactoryTest.groovy

```diff
@@ -2406,4 +2556,586 @@ schema {
+        def e = thrown(AbortExecutionException)
+        e.message == "Maximum field count exceeded. 2014 > 2013"
+    }
+
+    def "small query exceeding fields count"() {
+        String schema = """
+        type Query {
+            hello: String
+        }
+        """
+
+        GraphQLSchema graphQLSchema = TestUtil.schema(schema)
+
+        String query = """ {hello a1: hello}"""
+
+        assertValidQuery(graphQLSchema, query)
+
+        Document document = TestUtil.parseQuery(query)
+
+        when:
+        def result = ExecutableNormalizedOperationFactory.createExecutableNormalizedOperationWithRawVariables(
+                graphQLSchema,
+                document,
+                null,
+                RawVariables.emptyVariables(),
+                ExecutableNormalizedOperationFactory.Options.defaultOptions().maxFieldsCount(1))
+
+        then:
+        def e = thrown(AbortExecutionException)
+        e.message == "Maximum field count exceeded. 2 > 1"
+
+
+    }
+
+    def "query not exceeding fields count"() {
+        String schema = """
+        type Query {
+            dogs: [Dog]
+        }
+        type Dog {
+            name: String
+            breed: String
+        }
+        """
+
+        GraphQLSchema graphQLSchema = TestUtil.schema(schema)
+
+        String query = """ {dogs{name breed }}"""
+
+        assertValidQuery(graphQLSchema, query)
+
+        Document document = TestUtil.parseQuery(query)
+
+        when:
+        def result = ExecutableNormalizedOperationFactory.createExecutableNormalizedOperationWithRawVariables(
+                graphQLSchema,
+                document,
+                null,
+                RawVariables.emptyVariables(),
+                ExecutableNormalizedOperationFactory.Options.defaultOptions().maxFieldsCount(3))
+
+        then:
+        notThrown(AbortExecutionException)
+
+
+    }
+
+    def "query with meta fields exceeding fields count"() {
+        String schema = """
+        type Query {
+            hello: String
+        }
+        """
+
+        GraphQLSchema graphQLSchema = TestUtil.schema(schema)
+
+        String query = IntrospectionQuery.INTROSPECTION_QUERY
+
+        assertValidQuery(graphQLSchema, query)
+
+        Document document = TestUtil.parseQuery(query)
+
+        when:
+        def result = ExecutableNormalizedOperationFactory.createExecutableNormalizedOperationWithRawVariables(
+                graphQLSchema,
+                document,
+                null,
+                RawVariables.emptyVariables(),
+                // This limit is set to 188 in version 21+
+                ExecutableNormalizedOperationFactory.Options.defaultOptions().maxFieldsCount(187))
+        println result.normalizedFieldToMergedField.size()
+
+        then:
+        def e = thrown(AbortExecutionException)
+        // This line is different in version 21+, it is "Maximum field count exceeded. 189 > 188"
+        e.message == "Maximum field count exceeded. 188 > 187"
```

Comment:
The way we count nodes is 1 less in v20 compared to v21 & master

Reply:
Thanks Andi: the reason for this is `@oneOf` is introduced in v21, so v20 and prior will have 1 less node